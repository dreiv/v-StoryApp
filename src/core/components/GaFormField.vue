<script setup lang="ts">
import { type InputHTMLAttributes } from 'vue'

export interface FormFieldProps extends /* @vue-ignore */ InputHTMLAttributes {
  label?: string
}

defineProps<FormFieldProps>()
</script>

<template>
  <label :class="$style.field" v-bind="$attrs">
    <slot />

    <span :class="$style.label">
      <slot name="label">{{ label }}</slot>
    </span>
  </label>
</template>

<style module>
.field {
  display: inline-flex;
  position: relative;
  gap: var(--ga-size-spacing-03);

  user-select: none;
}

.label {
  font-weight: var(--ga-font-weight-normal);
  font-size: var(--ga-text-md-font-size);
  line-height: var(--ga-text-md-line-height);
  font-family: var(--ga-font-family-primary);
  letter-spacing: var(--ga-text-md-tracking);

  &:empty {
    display: none;
  }
}
</style>
