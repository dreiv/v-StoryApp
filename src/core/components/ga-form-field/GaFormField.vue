<script setup lang="ts">
import { computed, provide, useCssModule, reactive } from 'vue'
import type { vTooltip } from 'floating-vue'

import { uniqueId } from '@/core/utils/uniqueId'
import { formFieldKey, type FormFieldContext } from './types'

export interface FormFieldProps extends FormFieldContext {
  info?: string
  label?: string
  state?: string
  tooltip?: typeof vTooltip
}

const { id = uniqueId('form-field'), disabled, error } = defineProps<FormFieldProps>()

const style = useCssModule()
const classes = computed(() => [style.formField, { [style.disabled]: disabled }])

provide(
  formFieldKey,
  reactive({
    id,
    get disabled() {
      return disabled
    },
    get error() {
      return error
    },
  }),
)
</script>

<template>
  <div :class="classes">
    <label :class="$style.label" v-if="$slots.label || label" :for="id">
      <span :class="$style.text" v-tooltip="tooltip" :tabindex="tooltip && 0">
        <slot name="label">{{ label }}</slot>
      </span>

      <span :class="$style.state" v-if="$slots.state || state">
        <slot name="state">{{ state }}</slot>
      </span>
    </label>

    <slot />

    <div :class="$style.info" v-if="$slots.info || info">
      <slot name="info">{{ info }}</slot>
    </div>
  </div>
</template>

<style module>
.formField {
  display: flex;
  flex-direction: column;
  gap: var(--ga-size-spacing-03);

  &.disabled {
    color: var(--ga-color-text-disabled);
  }
}

.label {
  display: inline-flex;
  gap: var(--ga-size-spacing-02);
  border-radius: var(--ga-radius);
  padding-inline: var(--ga-size-spacing-02);

  width: fit-content;

  &:focus-within {
    outline: var(--ga-size-border-width-md) solid var(--ga-color-border-focus);
    outline-offset: var(--ga-size-spacing-01);
  }
}

.text {
  outline: none;
  font-weight: var(--ga-font-weight-medium);
  font-size: var(--ga-text-md-font-size);
  line-height: var(--ga-text-md-line-height);
}

.state {
  font-size: var(--ga-text-sm-font-size);
  line-height: var(--ga-text-sm-line-height);
}

.info {
  display: flex;
  align-items: center;
  gap: var(--ga-size-spacing-02);

  font-size: var(--ga-text-xs-font-size);
  line-height: var(--ga-text-xs-line-height);
  letter-spacing: var(--ga-text-xs-tracking);
}
</style>
