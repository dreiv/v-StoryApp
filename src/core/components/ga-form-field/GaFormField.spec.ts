import { shallowMount, type VueWrapper } from '@vue/test-utils'
import GaFormField from './GaFormField.vue'

describe('GaFormField', () => {
  let wrapper: VueWrapper

  it('should render correctly with default props and slot', () => {
    wrapper = shallowMount(GaFormField, {
      slots: { default: '<input type="text" />' },
    })
    expect(wrapper.html()).toMatchSnapshot()
    expect(wrapper.find('input').exists()).toBe(true)
  })

  it('should render label via prop', () => {
    wrapper = shallowMount(GaFormField, {
      props: { label: 'Test Label' },
      slots: { default: '<input type="text" />' },
    })
    expect(wrapper.text()).toContain('Test Label')
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should render label via slot', () => {
    wrapper = shallowMount(GaFormField, {
      slots: {
        default: '<input type="text" />',
        label: '<span class="custom-label">Custom Label</span>',
      },
    })
    expect(wrapper.find('.custom-label').exists()).toBe(true)
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should render info via prop', () => {
    wrapper = shallowMount(GaFormField, {
      props: { info: 'Test Info' },
      slots: { default: '<input type="text" />' },
    })
    expect(wrapper.text()).toContain('Test Info')
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should render info via slot', () => {
    wrapper = shallowMount(GaFormField, {
      slots: {
        default: '<input type="text" />',
        info: '<span class="custom-info">Custom Info</span>',
      },
    })
    expect(wrapper.find('.custom-info').exists()).toBe(true)
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should render state via prop', () => {
    wrapper = shallowMount(GaFormField, {
      props: { label: 'Test Label', state: 'Optional' },
      slots: { default: '<input type="text" />' },
    })
    expect(wrapper.text()).toContain('Optional')
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should render state via slot', () => {
    wrapper = shallowMount(GaFormField, {
      props: { label: 'Test Label' },
      slots: {
        default: '<input type="text" />',
        state: '<span class="custom-state">Custom State</span>',
      },
    })
    expect(wrapper.find('.custom-state').exists()).toBe(true)
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should apply disabled class when disabled prop is true', () => {
    wrapper = shallowMount(GaFormField, {
      props: { disabled: true },
      slots: { default: '<input type="text" />' },
    })
    expect(wrapper.classes()).toContain('disabled')
    expect(wrapper.html()).toMatchSnapshot()
  })
})
