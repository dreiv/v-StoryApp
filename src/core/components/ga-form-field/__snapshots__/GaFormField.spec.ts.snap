// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`GaFormField > should apply disabled class when disabled prop is true 1`] = `
"<div class="formField disabled">
  <!--v-if--><input type="text">
  <!--v-if-->
</div>"
`;

exports[`GaFormField > should render correctly with default props and slot 1`] = `
"<div class="formField">
  <!--v-if--><input type="text">
  <!--v-if-->
</div>"
`;

exports[`GaFormField > should render info via prop 1`] = `
"<div class="formField">
  <!--v-if--><input type="text">
  <div class="info">Test Info</div>
</div>"
`;

exports[`GaFormField > should render info via slot 1`] = `
"<div class="formField">
  <!--v-if--><input type="text">
  <div class="info"><span class="custom-info">Custom Info</span></div>
</div>"
`;

exports[`GaFormField > should render label via prop 1`] = `
"<div class="formField"><label class="label" for="form-field-2"><span class="text">Test Label</span>
    <!--v-if-->
  </label><input type="text">
  <!--v-if-->
</div>"
`;

exports[`GaFormField > should render label via slot 1`] = `
"<div class="formField"><label class="label" for="form-field-3"><span class="text"><span class="custom-label">Custom Label</span></span>
    <!--v-if-->
  </label><input type="text">
  <!--v-if-->
</div>"
`;

exports[`GaFormField > should render state via prop 1`] = `
"<div class="formField"><label class="label" for="form-field-6"><span class="text">Test Label</span><span class="state">Optional</span></label><input type="text">
  <!--v-if-->
</div>"
`;

exports[`GaFormField > should render state via slot 1`] = `
"<div class="formField"><label class="label" for="form-field-7"><span class="text">Test Label</span><span class="state"><span class="custom-state">Custom State</span></span></label><input type="text">
  <!--v-if-->
</div>"
`;
